File: backend/ocr.py
"""
Gemini OCR wrapper. Runs blocking model calls in threadpool from routes.
"""
from google import genai
from google.genai import types
from .genai_client import client

def _run_gemini_ocr(image_bytes: bytes, mime_type: str) -> str:
    """
    Blocking call to Gemini that returns the raw model output string.
    """
    # Create image part from bytes
    img_part = types.Part.from_bytes(data=image_bytes, mime_type=mime_type)
    
    prompt = """
Extract the textual contents of the provided image of handwritten notes and return STRICT JSON only.
Return your answer as a JSON object with exactly these two fields:
JSON format:
{
  "text": "<full extracted text as a single string>",
  "lines": ["line1", "line2", ...]
}
Both fields must always be present, even if empty. Do not add any commentary or extra fields. Return only valid JSON.
"""
    
    # Use the new Gemini API client
    response = client.models.generate_content(
        model="gemini-2.5-flash",
        contents=[prompt, img_part],
    )
    
    # return raw model text for downstream parsing and storage
    return response.text

File: backend/services/ocr.js
/**
 * Gemini OCR service for extracting text from images
 */
const { model } = require('./genaiClient');

/**
 * Runs OCR on an image using Gemini AI.
 * @param {Buffer} imageBuffer - The image data as a Buffer
 * @param {string} mimeType - The MIME type of the image
 * @returns {Promise<string>} The raw model output string
 */
async function runGeminiOCR(imageBuffer, mimeType) {
    const prompt = `
Extract the textual contents of the provided image of handwritten notes and return STRICT JSON only.
Return your answer as a JSON object with exactly these two fields:
JSON format:
{
  "text": "<full extracted text as a single string>",
  "lines": ["line1", "line2", ...]
}
Both fields must always be present, even if empty. Do not add any commentary or extra fields. Return only valid JSON.
`;

    // Convert buffer to base64
    const base64Image = imageBuffer.toString('base64');

    // Create the image part for the API
    const imagePart = {
        inlineData: {
            data: base64Image,
            mimeType: mimeType
        }
    };

    // Generate content with the model
    const result = await model.generateContent([prompt, imagePart]);
    const response = await result.response;

    return response.text();
}

module.exports = {
    runGeminiOCR
};
